### PR TITLE
[JENKINS-68373] Update form validation appearance

### PR DESCRIPTION
**Pull request by janfaracik**
https://user-images.githubusercontent.com/43062514/162643335-d5f31f12-d8d6-4eed-9809-3095a55bb0ec.mov

This PR updates the appearance of form validation, it also adds simple animations for it and simplifies the JS behind the scenes. 

### Proposed changelog entries

* Update form validation appearance.

### Proposed upgrade guidelines

N/A

### Proposed developer upgrade guidelines

* The markup for form warnings/errors has changed, if you're using custom logic around this area you may need to update your plugin accordingly

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by alecharp** [comment]
@janfaracik could you create a ticket on issues.jenkins.io? This is because we could have some regression on core or plugins and this would make it easier to track. Thanks.

**Comment by timja** [comment]
The stack trace expansion seems to go off the bottom of the page and looks funny:

image

I used the azure-credentials plugin and put garbage into the fields for creating a credential of type 'Azure Service Principal'.
(it takes a while to fail there's probably easier ways to get a stacktrace back from validation)

**Comment by janfaracik** [comment]
> @janfaracik could you create a ticket on issues.jenkins.io? This is because we could have some regression on core or plugins and this would make it easier to track. Thanks.

Done under https://issues.jenkins.io/browse/JENKINS-68373

**Comment by timja** [comment]
Hopefully that's all the tests fixed now 💪 

Also picked up a bug and corrected it in https://github.com/jenkinsci/jenkins/pull/6460/commits/2fc1775b66ee7e327fabc471e386bd75bc8ec622

**Comment by timja** [comment]
Seems at least one type of form validation hasn't been updated?

image

When configuring a view

Picked up in https://github.com/jenkinsci/acceptance-test-harness/blob/master/src/test/java/core/FormValidationTest.java#L42

**Comment by timja** [comment]
ATH passed https://github.com/jenkinsci/acceptance-test-harness/pull/765
Bom pending https://github.com/jenkinsci/bom/pull/1098 (should be okay looked like just a flake on the previous build)

**Review by NotMyFault — APPROVED** [review]
I really like that!

**Comment by basil** [comment]
What is the plan for triaging user-reported issues that stem from this (and other) significant UX change(s)? I raised concerns about past precedent with regard to issue triage in [this message](https://groups.google.com/g/jenkinsci-ux/c/AFe-kcqoEok/m/RglfPIkKAgAJ). At the time of this writing, these concerns remain unacknowledged.

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (timja) [inline war/src/main/webapp/scripts/hudson-behavior.js:530]: The change in https://github.com/jenkinsci/jenkins/pull/6460/commits/42cac629598d1f7f725a787aa178264a63e5622e broke expanding stacktraces.

A trivial way to reproduce is with a GitHub App credential from github-branch-source plugin, just put gibbish in the private key field

**Review by timja — CHANGES_REQUESTED** [review]

**Comment by timja** [comment]
> What is the plan for triaging user-reported issues that stem from this (and other) significant UX change(s)? I raised concerns about past precedent with regard to issue triage in [this message](https://groups.google.com/g/jenkinsci-ux/c/AFe-kcqoEok/m/RglfPIkKAgAJ). At the time of this writing, these concerns remain unacknowledged.

I've replied on the mailing list

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (janfaracik) [reply war/src/main/webapp/scripts/hudson-behavior.js:530]: Should be fixed now - thanks for spotting! :)

**Review by timja — APPROVED** [review]
LGTM, I've done a fair bit of manual testing and ATH and bom pass

File: core/src/main/resources/lib/form/entry.jelly
Comment (basil) [inline core/src/main/resources/lib/form/entry.jelly:94]: > LGTM, I've done a fair bit of manual testing and ATH and bom pass

ATH and BOM testing are [necessary but not sufficient](https://en.wikipedia.org/wiki/Necessity_and_sufficiency#Relationship_between_necessity_and_sufficiency) for a change of this scope. In order to fully ensure the change will not break existing consumers, we should ensure that there are no usages of APIs for which we are breaking compatibility.

The comment just a few lines above states:

> Needs to have a `tr` class otherwise custom uses of […] `findFollowingTR(listBox, "validation-error-area")` will break

This PR removes the `tr` class, which will cause usages of `findFollowingTR(listBox, "validation-error-area")` in plugins to break. I would advise scanning the plugin corpus looking for such usages, and if any are found in non-deprecated plugins, either:

- Deprecating the plugin in the Update Center, linking to this PR in the deprecation notice
- Filing a PR to make the plugin compatible with this PR and making a good faith effort to have the PR merged and released in a timely fashion

An example of this is `git-parameter-plugin`, which as of this PR now breaks with the following stack trace:

```
TypeError: findFollowingTR(...).firstChild is null
    gitParameterUpdateSelect http://127.0.0.1:8080/adjuncts/eb9dda3e/net/uaznia/lukanus/hudson/plugins/gitparameter/javascript/git-parameter-select.js:8
    <anonymous> http://127.0.0.1:8080/adjuncts/eb9dda3e/net/uaznia/lukanus/hudson/plugins/gitparameter/javascript/git-parameter-select.js:96
    h http://127.0.0.1:8080/static/eb9dda3e/scripts/hudson-behavior.js:1568
    refillOnChange http://127.0.0.1:8080/static/eb9dda3e/scripts/hudson-behavior.js:1583
    <anonymous> http://127.0.0.1:8080/adjuncts/eb9dda3e/net/uaznia/lukanus/hudson/plugins/gitparameter/javascript/git-parameter-select.js:94
    applySubtree http://127.0.0.1:8080/static/eb9dda3e/scripts/behavior.js:116
    applySubtree http://127.0.0.1:8080/static/eb9dda3e/scripts/behavior.js:111
    applySubtree http://127.0.0.1:8080/static/eb9dda3e/scripts/behavior.js:93
    apply http://127.0.0.1:8080/static/eb9dda3e/scripts/behavior.js:76
    start http://127.0.0.1:8080/static/eb9dda3e/scripts/behavior.js:71
    onload http://127.0.0.1:8080/static/eb9dda3e/scripts/behavior.js:133
    onload http://127.0.0.1:8080/static/eb9dda3e/scripts/behavior.js:133
    addLoadEvent http://127.0.0.1:8080/static/eb9dda3e/scripts/behavior.js:132
    createSearchBox http://127.0.0.1:8080/static/eb9dda3e/scripts/hudson-behavior.js:2049
    <anonymous> http://127.0.0.1:8080/job/freestyle/build?delay=0sec:9
behavior.js:119:31
```

Please note that although this is one example of a problem, it is not necessarily the only example. To make a change like this with confidence, the corpus of plugins should be scanned for usages of the relevant APIs and the results systematically examined. This observation applies not only to this PR but to any PR of this scope.

Comment (janfaracik) [reply core/src/main/resources/lib/form/entry.jelly:94]: Hey! Thanks for this.

Do you have any reproduction steps for the `git-parameter-plugin` breakage? I can't seem to get the error on my end.

I'll investigate the others and compile a list to fix or deprecate.

Thanks :)

Comment (basil) [reply core/src/main/resources/lib/form/entry.jelly:94]: I had never used this plugin before but I suspected that if I hit the `findFollowingTR` function that I would see a problem. Reading the code, it seemed to be executed when filling in parameters during a build, so I created a parameterized Freestyle job with a Git branch parameter and then tried to run that job. I got the error when I started typing stuff into the Git branch parameter field.

Comment (janfaracik) [reply core/src/main/resources/lib/form/entry.jelly:94]: Reproduced it and fixed it https://github.com/jenkinsci/git-parameter-plugin/pull/94 - thanks Basil!

Using this search https://github.com/search?q=org%3Ajenkinsci+findFollowingTR%28*%2C+%22validation-error-area%22%29&type=code it seems the only remaining plugin is https://github.com/jenkinsci/bmc-rpd-plugin, which hasn't been updated in six years, happy to update it or deprecate it if need be.

Thanks! :)

Comment (timja) [reply core/src/main/resources/lib/form/entry.jelly:94]: It’s basically unused and has an active security warning so we can leave that one

Comment (basil) [reply core/src/main/resources/lib/form/entry.jelly:94]: > Reproduced it and fixed it jenkinsci/git-parameter-plugin#94 - thanks Basil!

Great! Once that PR is merged and released the 59,936 users of Git Parameter can be assured of a smooth upgrade experience without regression.

> It’s basically unused and has an active security warning so we can leave that one

I agree that we can leave behind BMC RPD Plugin. In order to leave this plugin behind in a way that retains the trust of our users, one of whom [recently wrote](https://www.sghill.net/2022/de-risking-rollouts/) "the community does a wonderful job of being clear about what is not supported, and supporting well what is supported," I think we should add that plugin to the [deprecations list](https://github.com/jenkins-infra/update-center2/blob/master/resources/deprecations.properties). That way, it is clear to users that plain text credential storage is not the only reason that this plugin is not recommended.

Comment (janfaracik) [reply core/src/main/resources/lib/form/entry.jelly:94]: Opened a PR for that here - https://github.com/jenkins-infra/update-center2/pull/597 - hope I've done it correctly :) Cheers

File: core/src/main/resources/lib/form/entry.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/form/entry.jelly:89]: The comment on the two lines above the one I'm commenting on should be deleted, with https://github.com/jenkinsci/jenkins/pull/6460/files#r867286362 it makes no sense anymore.

**Comment by janfaracik** [comment]
So for an update on this PR:

✅ PR is feature complete/tests passing
☑️ Update git-parameter plugin to support API changes - https://github.com/jenkinsci/git-parameter-plugin/pull/94 (needs one more approval + need to merge)
☑️ Deprecate bmc-rpd plugin - https://github.com/jenkins-infra/update-center2/pull/597 (need to merge)

**Comment by klimas7** [comment]
Hi,
git-parameter plugin was released in version 0.9.17

**Review by basil — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (daniel-beck) [inline war/src/main/webapp/scripts/hudson-behavior.js:2347]: Several plugins used this, including LDAP plugin: https://github.com/search?q=user%3Ajenkinsci+applyErrorMessage&type=Code

**Comment by jglick** [comment]
This breaks form validation—errors are not properly cleared and replaced with new errors. (Noticed in https://github.com/jenkinsci/ldap-plugin/pull/185#issuecomment-1460028705.) Reproducible in 2.355 but not 2.354 from `/configureSecurity/` with a trivial plugin:

```java
import hudson.Extension;
import hudson.model.Descriptor;
import hudson.security.SecurityRealm;
import hudson.util.FormValidation;
import org.kohsuke.stapler.DataBoundConstructor;
import org.kohsuke.stapler.QueryParameter;
public final class XSecurityRealm extends SecurityRealm {
    public final String v;
    @DataBoundConstructor public XSecurityRealm(String v) {
        this.v = v;
    }
    @Override public SecurityComponents createSecurityComponents() {
        throw new UnsupportedOperationException();
    }
    @Extension public static final class DescriptorImpl extends Descriptor<SecurityRealm> {
        @Override public String getDisplayName() {
            return "X";
        }
        public FormValidation doCheckV(@QueryParameter String v) {
            return FormValidation.error(new IllegalStateException("Processing " + v), "Could not process " + v);
        }
    }
}
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?jelly escape-by-default='true'?>
<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
    <f:entry field="v" title="Some value">
        <f:textbox/>
    </f:entry>
</j:jelly>
```

**Comment by daniel-beck** [comment]
Filed the previous comment as [JENKINS-70793](https://issues.jenkins.io/browse/JENKINS-70793).

**Comment by yaroslavafenkin** [comment]
Also causes [JENKINS-71252](https://issues.jenkins.io/browse/JENKINS-71252)

**Comment by basil** [comment]
The following plugins remain broken by the removal of `applyErrorMessage`:

- https://plugins.jenkins.io/dingding-notifications/
- https://plugins.jenkins.io/JiraTestResultReporter/
- https://plugins.jenkins.io/oracle-cloud-infrastructure-compute/
- https://plugins.jenkins.io/wxwork-notification/

The most popular of these plugins has 5,801 installations.

**Pull request merged**
